### PR TITLE
SPIRVReader: Fix a padding issue of OpRuntimeArray

### PR DESCRIPTION
**Pull request by xuechen417**
When translating OpRuntimeArray in spir-v reader, there is an assumption
for required padding cases that the array stride should be greater than
the size of the array element. Howerver, there is an exception in HLSL
that the array element is a structure with variable-sized array as the
last member, which should not be identified as a padding case as the
behavior of GLSL.
To fix that, the assertion is removed and a new condition is added to
block this kind of arrays to be padded.

Fixes: assertion on dEQP-VK.tessellation.fractional_spacing.glsl_*

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by JaxLinAMD** [comment]
add to whitelist

**Comment by piotrAMD** [comment]
Can you please add a test inside "check-amdllpc" to demonstrate the corner case?

We are in the process of migrating to opaque pointers, and offline tests for corner cases would help dramatically.

Is the corner case related to the comment around line 585?

> // HLSL-derived shaders contain some (entirely valid) strange mappings for arrays that cannot be represented
>       // in LLVM. This manifests as an offset for a struct member that overlaps the previous data in the struct.
>       // To workaround this, we need to change the previous member in the struct to a pad array that we'll sort
>       // out during access-chain and load/stores later.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ff1cf6da17b6bc7d2160acc478202751035d4b22
<details><summary>CTS tests (Failed: 0/187823)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31111/57371 (54.2%)</li>
<li>Failed: 0/57371 (0.0%)</li>
<li>Not Supported: 26260/57371 (45.8%)</li><li>Warnings: 0/57371 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by xuechen417** [comment]
> Can you please add a test inside "check-amdllpc" to demonstrate the corner case?
> 
I'd like to add a lit test. But the assmebly text of the HLSL-derived domain shader cannot be correctly parsed. The use of BufferBlock operand required spir-v 1.3 or earlier. If I replace BufferBlock with Block, it will has other compiling errors. It seems the earlier spir-v version may cause the problem. Do you have any thoughts on it?
The domain shader looks like:
```
struct OutputStruct
{
    int numInvocations;
    float tessCoord[];
};
globallycoherent RWStructuredBuffer <OutputStruct> Output : register(b1);

void main(float2 tessCoords : SV_DOMAINLOCATION)
{
    int index;
    InterlockedAdd(Output[0].numInvocations, 1, index);
    Output[0].tessCoord[index] = tessCoords.x;
}
```
The related part of assembly text dumped from the CTS test:
```
...
OpDecorate %16 ArrayStride 4
OpMemberDecorate %17 0 Offset 0
OpMemberDecorate %17 1 Offset 4
OpDecorate %18 ArrayStride 8
OpMemberDecorate %19 0 Coherent
OpMemberDecorate %19 0 Offset 0
OpDecorate %19 BufferBlock
OpDecorate %21 DescriptorSet 0
OpDecorate %21 Binding 1
...
%16 = OpTypeRuntimeArray %6
%17 = OpTypeStruct %13 %16
%18 = OpTypeRuntimeArray %17
%19 = OpTypeStruct %18
%20 = OpTypePointer Uniform %19
%21 = OpVariable %20 Uniform
```

> We are in the process of migrating to opaque pointers, and offline tests for corner cases would help dramatically.
> 
> Is the corner case related to the comment around line 585?
> 
> > // HLSL-derived shaders contain some (entirely valid) strange mappings for arrays that cannot be represented
> > // in LLVM. This manifests as an offset for a struct member that overlaps the previous data in the struct.
> > // To workaround this, we need to change the previous member in the struct to a pad array that we'll sort
> > // out during access-chain and load/stores later.
No, the offset for this corner case is right.

**Comment by xuechen417** [comment]
V1: Add a lit test in check-amdllpc to check the corner case.

**Comment by amdvlk-admin** [comment]
### Test summary for commit dbe111a6c9eb514b71dc09ae996ce613c9b08cdc
<details><summary>CTS tests (Failed: 0/187823)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31111/57371 (54.2%)</li>
<li>Failed: 0/57371 (0.0%)</li>
<li>Not Supported: 26260/57371 (45.8%)</li><li>Warnings: 0/57371 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+;PipelineOptimizer
+;Vertex   Shader Profile Key: 0x1C9ABA8F9404FCF7FA84FA5F11B64FD4,
+;Tessellation control Shader Profile Key: 0x586E53D4AA4C92DF281677352C8270DD,
+;Tessellation evaluation Shader Profile Key: 0x82B4BACBF302B7567083656B51E09C6A,
+
+[CompileLog]
+(null) (size = 0 bytes)
+
+.AMDGPU.disasm (size = 5707 bytes)
+_amdgpu_gs_main:
```
Comment (amdrexu) [inline llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe:185]: Don't attach compile log (ISA, register config, binary code) in the .pipe file. It will make the file become longer unnecessarily.

File: llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe [hunk 2]
```diff
@@ -0,0 +1,430 @@
+; Check that the corner case that HLSL-derived shader that has a variable-sized array with struct-type element whose last member is variable-sized array.
+; The case can work well by not padding the outmost variable-sized array.
+
+; BEGIN_SHADERTEST
+; RUN: amdllpc -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST: AMDLLPC SUCCESS
```
Comment (amdrexu) [inline llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe:6]: Is it possible to add more pattern check in this test? Just check AMDLLPC SUCCESS is too simple. I prefer to add some IR detecting after translator or lower passes finish and output the text results.

File: llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe (hunk 1 above)
Comment (xuechen417) [reply llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe:185]: Thanks for your reminder. Fixed.

File: llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe (hunk 2 above)
Comment (xuechen417) [reply llpc/test/shaderdb/hlsl/PipelineHlsl_TestRuntimeArrayPacking.pipe:6]: Sure. I add a pattern by checking whether the global variable is padded.

**Review by xuechen417 — COMMENTED** [review]
V2: Remove useless info from the pipeline file and add a check pattern.

**Comment by amdvlk-admin** [comment]
### Test summary for commit a7729b78a712a6fba3ae5f973fa5285d8b446701
<details><summary>CTS tests (Failed: 0/187823)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31111/57371 (54.2%)</li>
<li>Failed: 0/57371 (0.0%)</li>
<li>Not Supported: 26260/57371 (45.8%)</li><li>Warnings: 0/57371 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by xuechen417** [comment]
V3: remove `-spvgen-dir=%spvgendir%` to fix the failures

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2ffaa225489052d32e17681f62a1777f319c3b3d
<details><summary>CTS tests (Failed: 1/174021)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26327/43569 (60.4%)</li>
<li><details><summary>Failed: 1/43569 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.ballot_broadcast.compute.subgroupbroadcast_i8vec4
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 17241/43569 (39.6%)</li><li>Warnings: 0/43569 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2ffaa225489052d32e17681f62a1777f319c3b3d
<details><summary>CTS tests (Failed: 0/186422)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 36645/65226 (56.2%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 28581/65226 (43.8%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30002/55970 (53.6%)</li>
<li>Failed: 0/55970 (0.0%)</li>
<li>Not Supported: 25968/55970 (46.4%)</li><li>Warnings: 0/55970 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37804/65226 (58.0%)</li>
<li>Failed: 0/65226 (0.0%)</li>
<li>Not Supported: 27422/65226 (42.0%)</li><li>Warnings: 0/65226 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Comment by kuhar** [comment]
@xuechen417 @amdrexu This PR added spir-v binary blobs as tests inputs. This is problematic because blobs are not versioned well by git and cannot be inspected during code review, etc., until someone decides to disassemble them.
Could you please change it to textual inputs (.spvasm)? We have examples of multi-input tests in shaderdb already.

**Comment by amdrexu** [comment]
@kuhar Sorry. It's my mistake. I didn't notice the pipe file links SPIR-V binary as input sources. They didn't show in review webpage. You are right. We should accept text input source for maintenance reason. Chen, Xue will make a follow-up change.

**Comment by xuechen417** [comment]
@kuhar Thank you for your comments. I made a PR #1811 to modify the test with textual inputs. Please review.
